### PR TITLE
Room Occupant may take a Seat at a Breakout Table

### PR DESCRIPTION
**Pull request by zspencer**
https://github.com/zinc-collective/convene/issues/154

### CODE REVIEW

**Comment by anaulin** [comment]
The direction this is going in makes sense to me, but it feels like it needs more to flesh out the fine detail of the interactions / data flow.

Quick thoughts after looking through the diff:
* What's the connection between `BreakoutRoom` and `Furniture`? I don't see any connections between the models in this PR, but I would expect `BreakoutRoom` to be a type of `Furniture` in some way?
* Related to the point above, I don't see changes to `schema.rb` -- seems like that is missing from the PR.
* Having a different partial for each type of `Furniture` to render it in the UI makes sense to me at first glance.
* How will we handle e.g. creating a BreakoutRoom? Is that a different furniture-type-specific partial to create/edit?

File: convene-web/app/views/rooms/show.html.erb [hunk 1]
```diff
@@ -13,6 +13,20 @@
 >
 </div>
 
+<%- room.furniture_placements.each do |furniture_placement| %>
+  <div
+    class="<%=furniture_placement.furniture_slug%>"
+    data-controller="<%=furniture_placement.furniture_slug %>"
+    <%- furniture_placement.data_tuples.each do |name, value| %>
```
Comment (anaulin) [inline convene-web/app/views/rooms/show.html.erb:20]: What are these `data_tuples`? Is that an intended field on `FurniturePlacement` that is currently missing from the model, or something that Rails magically provides? (Quick search for this didn't shed light, sadly.)

Comment (zspencer) [reply convene-web/app/views/rooms/show.html.erb:20]: My intent was for this to be a way to expose data to stimulus: https://stimulusjs.org/handbook/managing-state#the-data-api-explained. It's a bit of "coding by wishful thinking" in that I never actually defined them; but essentially they are a mechanism for a piece of furniture to expose data to the front-end.

**Comment by zspencer** [comment]
Re: What's the connection between BreakoutRoom and Furniture? I don't see any connections between the models in this PR, but I would expect BreakoutRoom to be a type of Furniture in some way?

I'm not sure yet;  my gut agrees with you that a BreakoutRoom is likely a _kind of_ furniture; and whether we go with Furniture as a Mixin type inheritance or Parent Child will depend on what makes sense when we have several types of furniture.

Re: Related to the point above, I don't see changes to schema.rb -- seems like that is missing from the PR.

Yep, I'm "coding by wishful thinking" here and sketching very lightly and not including some of the details that would make this code "work"; more trying to play with how the different objects interact.

Re: Having a different partial for each type of Furniture to render it in the UI makes sense to me at first glance.

:+1:

Re: How will we handle e.g. creating a BreakoutRoom? Is that a different furniture-type-specific partial to create/edit?

Hmmm, my preference would be for the piece of furniture to fully encapsulate it's data-model. I.e. the BreakoutRooms furniture would hold responsibility for deciding who is allowed to perform what CRUD operations on it's internal state based upon the rights the person has within the room. I need to noodle a bit more on how the interface for that would work; and a 0.1 version of BreakoutRooms may not have that degree of control.

File: convene-web/app/views/rooms/show.html.erb (hunk 1 above)
Comment (zspencer) [reply convene-web/app/views/rooms/show.html.erb]: I've ditched this in favor of each piece of Furniture being able to define it's own partial, so that we're not coupling implementation details of how we pass data to the UI too early..

**Comment by zspencer** [comment]
@anaulin / @user512 - Made a bit more progress on this, some of which probably wants to be it's own PR to reduce noise (Grouping the scopes with the fields they're querying in the room model; for instance...) 

There's also some work still to be done before I feel confident that it's worth trying to build a second kind of furniture with; but it feels like it's getting into a bit better shape...

Thoughts?

**Comment by anaulin** [comment]
@zspencer this direction makes sense to me, and it is nice to see it start to take shape!

To me it makes sense to get something like this checked in, and use it to get ourselves to play with "Tables" and iterating on them as needed. Later we can refine things to support more than one type of architecture, incorporating whatever we learn from "Tables".

(From a product standpoint, users are used to "Breakout Rooms" as terminology, so maybe we can call ours "Breakout Tables" or something like that? Otherwise I fear that if we call them externally "Tables" that would just be confusing to folks.)

**Comment by zspencer** [comment]
I'm a bit torn. I feel like "Breakout Rooms" breaks the [skeuomorphic mental model](https://blueprintdigital.com/skeuomorphic-design-the-pros-and-cons/) of the "Rent a Space, Lay it out with Rooms, add Furniture" framing that I'm using in the customer/contributor-development cycle.

That said, that attachment may be unproductive at the moment.

**Comment by user512** [comment]
I think table is fine and potentially is a good term if we foresee other furniture will be added in the future and I like the idea of getting this checked in and change if we have stronger feeling to change later.

> Realtime! When someone creates a breakout room; we want Guests to see that become available right away.

We can look into using [Turbo](https://turbo.hotwire.dev/), where someone created a table, the app would stream partial page updates over websocket. I haven't try using this library but by watching this [demo](https://gorails.com/episodes/hotwire-rails), it seems promising. 

> How do we kick people out of the breakout room when they expire?

Implement some JS to check if the room is expired every few minutes and redirect people if room is expired.

**Comment by anaulin** [comment]
I don't have super strong feelings on naming, but to clarify that above I was proposing something like "Breakout Tables", kinda splitting the difference (not "Breakout Rooms", as "Rooms" would be confusing in various ways in our model).

File: convene-web/app/lib/furniture/tables.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
+module Furniture
+  class Tables
```
Comment (user512) [inline convene-web/app/lib/furniture/tables.rb:2]: How do you feel about using plural for namespace and singular for model?
```ruby
module Furnitures
  class Tables
    ...
  end
end
```

**Comment by zspencer** [comment]
That makes sense, had a bit of a brain-fart! Updated to `BreakoutTables` and added in specs for the FurniturePlacement class; as well as fixed the breakages in the feature tests.

File: convene-web/app/models/furniture_placement.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+class FurniturePlacement < ApplicationRecord
+  belongs_to :room
+  delegate :workspace, to: :room
+
+  attribute :name, :string
+  validates :name, uniqueness: { scope: :room_id }
+
+  attribute :settings, :json
+
+  # TODO: Ordering of records is likely handled by a well-maintained gem already
```
Comment (user512) [inline convene-web/app/models/furniture_placement.rb:11]: Which gem were you referring to here?

File: convene-web/config/routes.rb [hunk 4]
```diff
@@ -1,26 +1,29 @@
...
-      root to: "workspaces#index"
-    end
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
   resources :workspaces do
-    resources :rooms, only: [:show, :edit, :update] do
-      resource :waiting_room, only: [:show, :update]
+    resources :rooms, only: %i[show edit update] do
+      resource :waiting_room, only: %i[show update]
+      namespace :furniture do
+        resources :tables, only: [:show], controller: 'breakout_tables/'
```
Comment (user512) [inline convene-web/config/routes.rb:20]: [Non-blocking] There is no show action in `BreakoutTables`, I assume this is intentional to be an example on how to add routes to point to this controller?

**Review by user512 — APPROVED** [review]
I think how furniture works is still in flux (like the shape of data in breakout table setting) but this seems safe to merge and developer can keep exploring how breakout table works.

File: convene-web/config/environments/test.rb [hunk 5]
```diff
@@ -12,7 +12,7 @@
   # Do not eager load code on boot. This avoids loading your whole application
   # just for the purpose of running a single test. If you are using a tool that
   # preloads Rails for running tests, you may have to set it to true.
-  config.eager_load = false
+  config.eager_load = true
```
Comment (user512) [inline convene-web/config/environments/test.rb:15]: Let me know if the solution in https://github.com/zinc-collective/convene/pull/172 looks good to you.
It can be merge as a separate PR or just merged into this branch.
